### ISSUE TITLE
Virtual root flag

### ISSUE THREAD

**Issue by scofalik** [ckeditor/ckeditor5-engine#479]
I just ran into a problem when I created another "graveyard-like" root. I needed it for `History` class to create "fake" operations needed for history compression. So I created the root and the problems started.

In `Selection`, if there are no ranges, we have `_getDefaultRange` method:

```javascript
	_getDefaultRange() {
		const defaultRoot = this._document._getDefaultRoot();

		// Find the first position where the selection can be put.
		for ( let position of Range.createFromElement( defaultRoot ).getPositions() ) {
			if ( this._document.schema.check( { name: '$text', inside: position } ) ) {
				return new Range( position, position );
			}
		}

		const position = new Position( defaultRoot, [ 0 ] );

		return new Range( position, position );
	}
```

It needs to find a proper place in "default root" `Document._getDefaultRoot`:

```javascript
	_getDefaultRoot() {
		for ( let root of this._roots.values() ) {
			if ( root !== this.graveyard ) {
				return root;
			}
		}

		return this.graveyard;
	}
```

Unfortunately, my "fake root" is added before "main" root so it is returned. But there is no data there, so, later on, mappings fail and crash the editor.

I open to this issue to have a discussion whether we allow other virtual roots like "graveyard" and if so, how will we differentiate them from "real" roots. Proposals are: virtual roots starts from `$` or just flag in `RootElement`.

**Comment by Reinmar** [comment]
> Proposals are: virtual roots starts from $ or just flag in RootElement.

Flag may be better, because we already use `$` in e.g. `$text` or `$root` names in the schema, but hence also in the document element names...

**Comment by pjasiun** [comment]
> > Proposals are: virtual roots starts from $ or just flag in RootElement.
>
> Flag may be better, because we already use $ in e.g. $text or $root names in the schema, but hence also in the document element names...

On the other hand `$` always mean that it is a "special" element. We already have `$graveyard` so in make sense that `_getDefaultRange` returns the range in the first root which is not `$*`.

Also, since we introduced `main` root I do not see much more usage for `_getDefaultRoot` then the one in the selection. I think that selection could simple use `getRoot()`. Then, for the multi root support we could introduce `defaultRootName` property which is `main` by default and this is the name of the root which should be return by `getRoot()` (with no `name` parameter). This property can be overwritten, so if one create a multi-root editor, and will have no `main` root, he can overwrite this property. But, for now, we could assume that there is always a `main` root, even if there are multiple roots.

**TL;DR:** `_getDefaultRoot` and `getRoot` do the same thing, there is no point to have both, use `getRoot`.

**Comment by Reinmar** [comment]
In my opinion these are two different things. `getRoot()` returns you a root by name, where name defaults to `'main'`. If you have multi-root editor, then you may not have the main one. And even if you had like title+main fields, the main one should not be the default root. Why? Because the first focusable field would be the title.

So for the selection the order of roots matters. It should get the first normal root and this is more or less what the `getDefaultRange()` tries to do now.

Hence, I would keep something like `getDefaultRoot()` or `getFirstRoot()`. Such a method should return the first normal root.

**Comment by pjasiun** [comment]
> In my opinion these are two different things. getRoot() returns you a root by name, where name defaults to 'main'. If you have multi-root editor, then you may not have the main one. And even if you had like title+main fields, the main one should not be the default root. Why? Because the first focusable field would be the title.

You are right.

> So for the selection the order of roots matters. It should get the first normal root and this is more or less what the getDefaultRange() tries to do now.
> 
> Hence, I would keep something like getDefaultRoot() or getFirstRoot(). Such a method should return the first normal root.

If the default selection position is the only case we can come up, we should create such method on the selection. It will be clear what such method is for. Having both `getRoot()` and `getDefaultRoot()` on the document will be misleading.
